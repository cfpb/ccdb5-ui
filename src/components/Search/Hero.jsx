import './Hero.less';
import { MODAL_SHOWN, MODAL_TYPE_MORE_ABOUT } from '../../constants'
import { connect } from 'react-redux'
import React from 'react';

/**
* Notifies the application that the "More About..." dialog box should appear
*
* @returns {string} a packaged payload to be used by Redux reducers
*/
export function showMoreAboutDialog() {
  return {
    type: MODAL_SHOWN,
    modalType: MODAL_TYPE_MORE_ABOUT,
    modalProps: {}
  }
}

export class Hero extends React.Component {
  render() {
    return (
        <header className="content_hero">
          <h1 className="content-header">Consumer Complaint Database</h1>
          <ul className="m-list m-list__horizontal">
            <li className="m-list_item">
              <button className="a-btn a-btn__link"
                      onClick={this.props.onMoreAbout}>
                Things to know before you use this database
              </button>
            </li>
            <li className="m-list_item">
              <a href="https://www.consumerfinance.gov/complaint/data-use/"
                 target="_blank"
                 rel="noopener noreferrer">
                How we use complaint data
              </a>
            </li>
            <li className="m-list_item">
              <a href="https://cfpb.github.io/api/ccdb/"
                 target="_blank"
                 rel="noopener noreferrer">
                Technical documentation
              </a>
            </li>
          </ul>
        </header>
    );
  }
}

export const mapDispatchToProps = dispatch => ( {
  onMoreAbout: () => {
    dispatch( showMoreAboutDialog() )
  }
} )

export default connect( null, mapDispatchToProps )( Hero )
