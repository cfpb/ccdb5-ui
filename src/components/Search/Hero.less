@import (less) '../../css/base.less';

.content_hero {
  padding: @gutter-normal;

  @media @desktop {
    padding: @grid_gutter-width;
    padding-top: @gutter-normal;
  }

  @media @phone {
    padding: @gutter-narrow;
  }

  margin-top: @grid_gutter-width;

  p {
    max-width: 670px;
  }

  ul {
    padding-top: @gutter-minimum;
  }

  li {
    padding-bottom: 0;
    border-bottom: 0;

    button, a {
      font-size: 16px;
    }

    &.m-list_item {
      margin-right: 0;
    }

    &::after {
      margin-left: @gutter-narrow;
      margin-right: @gutter-narrow;

      content: '|';
    }

    &:last-of-type {
      &::after {
        margin: 0;

        content: '';
      }
    }
  }

  .content-header {
    display:inline-flex;
  }
  h4.content-header {
    box-sizing: border-box;
    margin-left: @gutter-narrow;
    margin-top: 0;
    color: @dark-gray;
    text-transform: uppercase;
    vertical-align: text-bottom;
  }
}
