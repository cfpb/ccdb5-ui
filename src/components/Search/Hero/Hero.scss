@use 'sass:math';
@use '@cfpb/cfpb-design-system/src/abstracts' as *;
@use 'css/base' as *;

.content__hero {
  padding-bottom: math.div(30px, $base-font-size-px) + rem;

  p {
    max-width: 670px;
  }

  ul {
    padding-top: $gutter-minimum;
  }

  li {
    padding-bottom: 0;
    border-bottom: 0;

    button,
    a {
      font-size: 16px;
      display: inline-block;
    }

    &.m-list__item {
      margin-right: 0;
    }

    &::after {
      margin-left: $gutter-narrow;
      margin-right: $gutter-narrow;

      content: '|';
    }

    &:last-of-type {
      &::after {
        margin: 0;

        content: '';
      }
    }
  }

  .content-header {
    display: inline-flex;
  }
  h4.content-header {
    box-sizing: border-box;
    margin-left: $gutter-narrow;
    margin-top: 0;
    color: var(--gray-dark);
    text-transform: uppercase;
    vertical-align: text-bottom;
  }
}
