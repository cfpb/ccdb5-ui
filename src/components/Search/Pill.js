import './Pill.less';
import {
  removeFilter,
  changeDateRange,
  replaceFilters,
} from '../../reducers/query/query';
import { filterPatch, SLUG_SEPARATOR } from '../../constants';
import { formatPillPrefix, getUpdatedFilters } from '../../utils/filters';
import { useDispatch, useSelector } from 'react-redux';
import { coalesce } from '../../utils';
import iconMap from '../iconMap';
import PropTypes from 'prop-types';
import React from 'react';
import { selectAggsState } from '../../reducers/aggs/selectors';
import { selectQueryState } from '../../reducers/query/selectors';

export const Pill = ({ fieldName, value }) => {
  const aggsState = useSelector(selectAggsState);
  const queryState = useSelector(selectQueryState);
  const aggs = coalesce(aggsState, fieldName, []);
  const filters = coalesce(queryState, fieldName, []);
  const prefix = formatPillPrefix(fieldName);
  const trimmed = value.split(SLUG_SEPARATOR).pop();
  const dispatch = useDispatch();

  const remove = () => {
    if (fieldName === 'date_received') {
      // reset date range
      dispatch(changeDateRange('All'));
    } else {
      const filterName = value;
      if (filterPatch.includes(fieldName)) {
        const updatedFilters = getUpdatedFilters(
          filterName,
          filters,
          aggs,
          fieldName
        );
        dispatch(replaceFilters(fieldName, updatedFilters));
      } else {
        dispatch(
          removeFilter({ fieldName: fieldName, fieldValue: filterName })
        );
      }
    }
  };

  return (
    <li>
      <button className="pill flex-fixed" onClick={remove}>
        <span className="name">
          {prefix}
          {trimmed}
        </span>
        <span className="u-visually-hidden">
          {`Remove ${trimmed} as a filter`}
        </span>
        {iconMap.getIcon('delete')}
      </button>
    </li>
  );
};

Pill.propTypes = {
  fieldName: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
};
