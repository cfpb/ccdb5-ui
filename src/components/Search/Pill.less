.pill {
  padding: @grid_gutter-width / 3 30px @grid_gutter-width / 3 @grid_gutter-width /
    3;
  border: 1px solid @teal-60;
  margin-right: 4px;
  border-radius: 0.28571429em;
  background-color: @teal-40;
  font-weight: 500;
  color: @black;
  font-size: @size-v;
  &:hover {
    background-color: #c4dddc;
    border-color: #9ec4c3;
    cursor: pointer;
  }
  // Close button
  button {
    padding: 0;
    padding-left: 5px;
    border: 0;
    border-left: 1px solid @gray-20;
    margin: 0;
    background-color: transparent;
    color: @teal-dark;
    position: absolute;
    right: 10px;
    top: 50%;
    transform: translateY(-50%);
    &:hover,
    &:active {
      background-color: @gray-40;
    }
  }
}
