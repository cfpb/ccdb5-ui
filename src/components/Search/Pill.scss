@use 'sass:math';
@use '@cfpb/cfpb-design-system/src/abstracts' as *;

.pill {
  padding: math.div($grid-gutter-width, 3) 30px math.div($grid-gutter-width, 3) math.div($grid-gutter-width, 3);
  background-color: var(--teal-20);
  border: 1px solid var(--teal-40);
  margin-right: 4px;
  border-radius: 0.28571429em;
  font-weight: 500;
  color: var(--black);
  font-size: $size-v;
  &:hover {
    background-color: var(--teal-30);
    border-color: var(--teal-50);
    cursor: pointer;
  }

  // Close button
  button,
  .cf-icon-svg {
    padding: 0;
    padding-left: 5px;
    border: 0;
    margin: 0;
    background-color: transparent;
    color: var(--teal-80);
    position: absolute;
    right: 15px;
    top: 50%;
    transform: translateY(-50%);
  }
}
