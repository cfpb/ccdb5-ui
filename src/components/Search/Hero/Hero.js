import './Hero.scss';
import { useDispatch } from 'react-redux';
import { LINK_DATA_USE, MODAL_TYPE_MORE_ABOUT } from '../../../constants';
import { modalShown } from '../../../reducers/view/viewSlice';

export const Hero = () => {
  const dispatch = useDispatch();

  return (
    <header className="content__hero">
      <h1 className="content-header">Consumer Complaint Database</h1>
      <ul className="m-list m-list--horizontal">
        <li className="m-list__item">
          <button
            className="a-btn a-btn--link"
            onClick={() => {
              dispatch(modalShown(MODAL_TYPE_MORE_ABOUT));
            }}
          >
            Things to know before you use this database
          </button>
        </li>
        <li className="m-list__item">
          <a href={LINK_DATA_USE} target="_blank" rel="noopener noreferrer">
            How we use complaint data
          </a>
        </li>
        <li className="m-list__item">
          <a
            href="https://cfpb.github.io/api/ccdb/"
            target="_blank"
            rel="noopener noreferrer"
          >
            Technical documentation
          </a>
        </li>
      </ul>
    </header>
  );
};
