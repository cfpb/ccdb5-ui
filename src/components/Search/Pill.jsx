import './Pill.scss';
import { dateRangeChanged } from '../../reducers/query/querySlice';
import {
  filterRemoved,
  filtersReplaced,
} from '../../reducers/filters/filtersSlice';
import { filterPatch, SLUG_SEPARATOR } from '../../constants';
import { formatPillPrefix, getUpdatedFilters } from '../../utils/filters';
import { useDispatch, useSelector } from 'react-redux';
import { coalesce } from '../../utils';
import getIcon from '../Common/Icon/iconMap';
import PropTypes from 'prop-types';
import { selectFiltersRoot } from '../../reducers/filters/selectors';
import { useGetAggregations } from '../../api/hooks/useGetAggregations';

export const Pill = ({ fieldName, value }) => {
  const { data: aggsState } = useGetAggregations();
  const filtersState = useSelector(selectFiltersRoot);
  const aggs = coalesce(aggsState, fieldName, []);
  const filters = coalesce(filtersState, fieldName, []);
  const prefix = formatPillPrefix(fieldName);
  const trimmed = value.split(SLUG_SEPARATOR).pop();
  const dispatch = useDispatch();

  const remove = () => {
    if (fieldName === 'date_received') {
      // reset date range
      dispatch(dateRangeChanged('All'));
    } else {
      const filterName = value;
      if (filterPatch.includes(fieldName)) {
        const updatedFilters = getUpdatedFilters(
          filterName,
          filters,
          aggs,
          fieldName,
        );
        dispatch(filtersReplaced(fieldName, updatedFilters));
      } else {
        dispatch(filterRemoved(fieldName, filterName));
      }
    }
  };

  return (
    <li>
      <button className="pill flex-fixed" onClick={remove}>
        <span className="name">
          {prefix}
          {trimmed}
        </span>
        {getIcon('delete')}
      </button>
    </li>
  );
};

Pill.propTypes = {
  fieldName: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
};
