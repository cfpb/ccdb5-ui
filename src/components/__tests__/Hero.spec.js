import { mapDispatchToProps } from '../Search/Hero';

describe('mapDispatchToProps', () => {
  it('hooks into showMoreAboutDialog', () => {
    const dispatch = jest.fn();
    mapDispatchToProps(dispatch).onMoreAbout();
    expect(dispatch.mock.calls.length).toEqual(1);
  });
});
