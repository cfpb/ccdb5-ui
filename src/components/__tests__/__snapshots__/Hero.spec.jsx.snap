// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initial state renders without crashing 1`] = `
<header
  className="content_hero"
>
  <h1
    className="content-header"
  >
    Consumer Complaint Database
  </h1>
  <ul
    className="m-list m-list__horizontal"
  >
    <li
      className="m-list_item"
    >
      <button
        className="a-btn a-btn__link"
        onClick={[Function]}
      >
        Things to know before you use this database
      </button>
    </li>
    <li
      className="m-list_item"
    >
      <a
        href="https://www.consumerfinance.gov/complaint/data-use/"
        rel="noopener noreferrer"
        target="_blank"
      >
        How we use complaint data
      </a>
    </li>
    <li
      className="m-list_item"
    >
      <a
        href="https://cfpb.github.io/api/ccdb/"
        rel="noopener noreferrer"
        target="_blank"
      >
        Technical documentation
      </a>
    </li>
  </ul>
</header>
`;
